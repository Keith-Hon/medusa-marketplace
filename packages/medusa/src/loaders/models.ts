import formatRegistrationName from "../utils/format-registration-name"
import glob from "glob"
import path from "path"
import { ClassConstructor, MedusaContainer } from "../types/global"
import { EntitySchema } from "typeorm"
import { asClass, asValue, AwilixContainer } from "awilix"

/**
 * Registers all models in the model directory
 */
export default ({ container }: { container: MedusaContainer }, config = { register: true }) => {
  const corePath = "../models/*.js"
  const coreFull = path.join(__dirname, corePath)

  const models: (ClassConstructor<unknown> | EntitySchema)[] = []

  const core = glob.sync(coreFull, { cwd: __dirname, ignore: ["index.js", "index.ts"] })
  core.forEach((fn) => {
    const loaded = require(fn) as ClassConstructor<unknown> | EntitySchema
    if (loaded) {
     Object.entries(loaded).map(([, val]: [string, ClassConstructor<unknown> | EntitySchema]) => {
        if (typeof val === "function" || val instanceof EntitySchema) {
          if (config.register) {
            const name = formatRegistrationName(fn)
            container.register({
              [name]: asClass(val as ClassConstructor<unknown>),
            })

            container.registerAdd("db_entities", asValue(val))
          }

          models.push(val)
        }
      })
    }
  })

  return models
}
