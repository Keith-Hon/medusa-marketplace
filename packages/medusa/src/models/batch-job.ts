import {
  AfterLoad,
  BeforeInsert,
  Column,
  Entity,
  JoinColumn,
  ManyToOne,
} from "typeorm"
import {
  BatchJobResultError,
  BatchJobResultStatDescriptor,
  BatchJobStatus,
} from "../types/batch-job"
import { DbAwareColumn, resolveDbType } from "../utils/db-aware-column"
import { SoftDeletableEntity } from "../interfaces/models/soft-deletable-entity"
import { generateEntityId } from "../utils/generate-entity-id"
import { User } from "./user"

@Entity()
export class BatchJob extends SoftDeletableEntity {
  @DbAwareColumn({ type: "text" })
  type: string

  @Column({ nullable: true })
  created_by: string | null

  @ManyToOne(() => User)
  @JoinColumn({ name: "created_by" })
  created_by_user: User

  @DbAwareColumn({ type: "jsonb", nullable: true })
  context: Record<string, unknown>

  @DbAwareColumn({ type: "jsonb", nullable: true })
  result: {
    count?: number
    advancement_count?: number
    progress?: number
    errors?: BatchJobResultError[]
    stat_descriptors?: BatchJobResultStatDescriptor[]
    file_key?: string
    file_size?: number
  } & Record<string, unknown>

  @Column({ type: "boolean", nullable: false, default: false })
  dry_run = false

  @Column({ type: resolveDbType("timestamptz"), nullable: true })
  pre_processed_at?: Date

  @Column({ type: resolveDbType("timestamptz"), nullable: true })
  processing_at?: Date

  @Column({ type: resolveDbType("timestamptz"), nullable: true })
  confirmed_at?: Date

  @Column({ type: resolveDbType("timestamptz"), nullable: true })
  completed_at?: Date

  @Column({ type: resolveDbType("timestamptz"), nullable: true })
  canceled_at?: Date

  @Column({ type: resolveDbType("timestamptz"), nullable: true })
  failed_at?: Date

  status: BatchJobStatus

  @AfterLoad()
  loadStatus(): void {
    /* Always keep the status order consistent. */
    if (this.pre_processed_at) {
      this.status = BatchJobStatus.PRE_PROCESSED
    }
    if (this.confirmed_at) {
      this.status = BatchJobStatus.CONFIRMED
    }
    if (this.processing_at) {
      this.status = BatchJobStatus.PROCESSING
    }
    if (this.completed_at) {
      this.status = BatchJobStatus.COMPLETED
    }
    if (this.canceled_at) {
      this.status = BatchJobStatus.CANCELED
    }
    if (this.failed_at) {
      this.status = BatchJobStatus.FAILED
    }

    this.status = this.status ?? BatchJobStatus.CREATED
  }

  @BeforeInsert()
  private beforeInsert(): void {
    this.id = generateEntityId(this.id, "batch")
  }

  toJSON() {
    this.loadStatus()
    return this
  }
}

/**
 * @schema batch_job
 * title: "Batch Job"
 * description: "A Batch Job."
 * x-resourceId: batch_job
 * properties:
 *  id:
 *    description: "The unique identifier for the batch job."
 *    type: string
 *  type:
 *    description: "The type of batch job."
 *    type: string
 *    enum:
 *      - product_import
 *      - product_export
 *  status:
 *    description: "The status of the batch job."
 *    type: string
 *    enum:
 *      - created
 *      - pre_processed
 *      - processing
 *      - completed
 *      - canceled
 *      - failed
 *  created_by:
 *    description: "The unique identifier of the user that created the batch job."
 *    type: string
 *  context:
 *    description: "The context of the batch job, the type of the batch job determines what the context should contain."
 *    type: object
 *  dry_run:
 *    description: "Specify if the job must apply the modifications or not."
 *    type: boolean
 *    default: false
 *  result:
 *    description: "The result of the batch job."
 *    type: object
 *  pre_processed_at:
 *    description: "The date from which the job has been pre processed."
 *    type: string
 *    format: date-time
 *  confirmed_at:
 *    description: "The date when the confirmation has been done."
 *    type: string
 *    format: date-time
 *  completed_at:
 *    description: "The date of the completion."
 *    type: string
 *    format: date-time
 *  canceled_at:
 *    description: "The date of the concellation."
 *    type: string
 *    format: date-time
 *  created_at:
 *    description: "The date with timezone at which the resource was created."
 *    type: string
 *    format: date-time
 *  updated_at:
 *    description: "The date with timezone at which the resource was last updated."
 *    type: string
 *    format: date-time
 *  deleted_at:
 *    description: "The date with timezone at which the resource was deleted."
 *    type: string
 *    format: date-time
 */
