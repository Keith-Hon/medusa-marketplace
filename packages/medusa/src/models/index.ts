export * from "./address"
export * from "./batch-job"
export * from "./cart"
export * from "./claim-image"
export * from "./claim-item"
export * from "./claim-order"
export * from "./claim-tag"
export * from "./country"
export * from "./currency"
export * from "./custom-shipping-option"
export * from "./customer"
export * from "./customer-group"
export * from "./discount"
export * from "./discount-condition"
export * from "./discount-condition-customer-group"
export * from "./discount-condition-product"
export * from "./discount-condition-product-collection"
export * from "./discount-condition-product-tag"
export * from "./discount-condition-product-type"
export * from "./discount-rule"
export * from "./draft-order"
export * from "./fulfillment"
export * from "./fulfillment-item"
export * from "./fulfillment-provider"
export * from "./gift-card"
export * from "./gift-card-transaction"
export * from "./idempotency-key"
export * from "./image"
export * from "./invite"
export * from "./line-item"
export * from "./line-item-adjustment"
export * from "./line-item-tax-line"
export * from "./money-amount"
export * from "./note"
export * from "./notification"
export * from "./oauth"
export * from "./order"
export * from "./payment"
export * from "./payment-provider"
export * from "./payment-session"
export * from "./price-list"
export * from "./product"
export * from "./product-collection"
export * from "./product-option"
export * from "./product-option-value"
export * from "./product-tag"
export * from "./product-tax-rate"
export * from "./product-type"
export * from "./product-type-tax-rate"
export * from "./product-variant"
export * from "./refund"
export * from "./region"
export * from "./return"
export * from "./return-item"
export * from "./return-reason"
export * from "./sales-channel"
export * from "./shipping-method"
export * from "./shipping-method-tax-line"
export * from "./shipping-option"
export * from "./shipping-option-requirement"
export * from "./shipping-profile"
export * from "./shipping-tax-rate"
export * from "./staged-job"
export * from "./store"
export * from "./swap"
export * from "./tax-provider"
export * from "./tax-rate"
export * from "./tracking-link"
export * from "./user"
