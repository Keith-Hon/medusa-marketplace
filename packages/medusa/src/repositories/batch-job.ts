import { EntityRepository, Repository } from "typeorm"
import { BatchJob } from "../models"

@EntityRepository(BatchJob)
export class BatchJobRepository extends Repository<BatchJob> {}
