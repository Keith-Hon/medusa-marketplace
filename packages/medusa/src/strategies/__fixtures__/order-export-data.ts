import { DeepPartial } from "typeorm"
import { IdMap } from "medusa-test-utils"
import {
  FulfillmentStatus,
  Order,
  OrderStatus,
  PaymentStatus,
} from "../../models"

const createdAtDate = new Date("2019-01-01T00:00:00.000Z")

export const ordersToExport: DeepPartial<Order>[] = [
  {
    id: "order_1",
    created_at: createdAtDate,
    display_id: 123,
    status: OrderStatus.PENDING,
    fulfillment_status: FulfillmentStatus.PARTIALLY_FULFILLED,
    payment_status: PaymentStatus.CAPTURED,
    subtotal: 10,
    shipping_total: 10,
    discount_total: 0,
    gift_card_total: 0,
    refunded_total: 0,
    tax_total: 5,
    total: 25,
    currency_code: "usd",
    region_id: "region_1",
    shipping_address: {
      id: "address_1",
      address_1: "123 Main St",
      address_2: "",
      city: "New York",
      country_code: "US",
      postal_code: "10001",
    },
    customer: {
      id: "customer_1",
      first_name: "John",
      last_name: "Doe",
      email: "John@Doe.com",
    },
    sales_channel: {
      id: IdMap.getId("sc_1"),
      name: "SC 1",
      description: "SC 1",
    },
  },
  {
    id: "order_2",
    created_at: createdAtDate,
    display_id: 124,
    status: OrderStatus.COMPLETED,
    fulfillment_status: FulfillmentStatus.FULFILLED,
    payment_status: PaymentStatus.CAPTURED,
    subtotal: 125,
    shipping_total: 10,
    discount_total: 0,
    gift_card_total: 0,
    refunded_total: 0,
    tax_total: 0,
    total: 135,
    currency_code: "eur",
    region_id: "region_2",
    shipping_address: {
      id: "address_2",
      address_1: "Hovedgaden 1",
      address_2: "",
      city: "Copenhagen",
      country_code: "DK",
      postal_code: "1150",
    },
    customer: {
      id: "customer_2",
      first_name: "Jane",
      last_name: "Doe",
      email: "Jane@Doe.com",
    },
    sales_channel: {
      id: IdMap.getId("sc_2"),
      name: "SC 2",
      description: "SC 2",
    },
  },
]
