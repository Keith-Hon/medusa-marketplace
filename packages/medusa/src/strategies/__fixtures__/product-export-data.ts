import { IdMap } from "medusa-test-utils"

const productIds = [
  "product-export-strategy-product-1",
  "product-export-strategy-product-2",
]
const variantIds = [
  "product-export-strategy-variant-1",
  "product-export-strategy-variant-2",
  "product-export-strategy-variant-3",
]
export const productsToExport = [
  {
    sales_channels: [
      { id: IdMap.getId("sc_1"), name: "SC 1", description: "SC 1" },
    ],
    collection: {
      created_at: "randomString",
      deleted_at: null,
      handle: "test-collection1",
      id: IdMap.getId("product-export-collection_1"),
      metadata: null,
      title: "Test collection 1",
      updated_at: "randomString",
    },
    collection_id: IdMap.getId("product-export-collection_1"),
    created_at: "randomString",
    deleted_at: null,
    description: "test-product-description-1",
    discountable: true,
    external_id: null,
    handle: "test-product-product-1",
    height: null,
    hs_code: null,
    id: productIds[0],
    images: [
      {
        created_at: "randomString",
        deleted_at: null,
        id: IdMap.getId("product-export-image_1"),
        metadata: null,
        updated_at: "randomString",
        url: "test-image.png",
      },
    ],
    is_giftcard: false,
    length: null,
    material: null,
    metadata: null,
    mid_code: null,
    options: [
      {
        created_at: "randomString",
        deleted_at: null,
        id: IdMap.getId("product-export-option_1"),
        metadata: null,
        product_id: productIds[0],
        title: "test-option-1",
        updated_at: "randomString",
      },
      {
        created_at: "randomString2",
        deleted_at: null,
        id: IdMap.getId("product-export-option_2"),
        metadata: null,
        product_id: productIds[0],
        title: "test-option-2",
        updated_at: "randomString2",
      },
    ],
    origin_country: null,
    profile_id: IdMap.getId("product-export-profile_1"),
    profile: {
      id: IdMap.getId("product-export-profile_1"),
      name: "profile_1",
      type: "profile_type_1",
    },
    status: "draft",
    subtitle: null,
    tags: [
      {
        created_at: "randomString",
        deleted_at: null,
        id: IdMap.getId("product-export-tag_1"),
        metadata: null,
        updated_at: "randomString",
        value: "123_1",
      },
    ],
    thumbnail: null,
    title: "Test product",
    type: {
      created_at: "randomString",
      deleted_at: null,
      id: IdMap.getId("product-export-type_1"),
      metadata: null,
      updated_at: "randomString",
      value: "test-type-1",
    },
    type_id: IdMap.getId("product-export-type_1"),
    updated_at: "randomString",
    variants: [
      {
        allow_backorder: false,
        barcode: "test-barcode",
        calculated_price: null,
        created_at: "randomString",
        deleted_at: null,
        ean: "test-ean",
        height: null,
        hs_code: null,
        id: variantIds[0],
        inventory_quantity: 10,
        length: null,
        manage_inventory: true,
        material: null,
        metadata: null,
        mid_code: null,
        options: [
          {
            created_at: "randomString",
            deleted_at: null,
            id: IdMap.getId("product-export-variant_option_1"),
            metadata: null,
            option_id: IdMap.getId("product-export-option_1"),
            updated_at: "randomString",
            value: "option 1 value 1",
            variant_id: variantIds[0],
          },
          {
            created_at: "randomString",
            deleted_at: null,
            id: IdMap.getId("product-export-variant_option_2"),
            metadata: null,
            option_id: IdMap.getId("product-export-option_2"),
            updated_at: "randomString",
            value: "option 2 value 1",
            variant_id: variantIds[0],
          },
        ],
        origin_country: null,
        original_price: null,
        prices: [
          {
            amount: 100,
            created_at: "randomString",
            deleted_at: null,
            id: IdMap.getId("product-export-price_1"),
            region_id: IdMap.getId("product-export-region_1"),
            max_quantity: null,
            min_quantity: null,
            price_list: null,
            price_list_id: null,
            region: {
              id: IdMap.getId("product-export-region_1"),
              currency_code: "usd",
              name: "france",
            },
            updated_at: "randomString",
            variant_id: variantIds[0],
          },
          {
            amount: 110,
            created_at: "randomString",
            currency_code: "usd",
            deleted_at: null,
            id: IdMap.getId("product-export-price_1"),
            region_id: null,
            max_quantity: null,
            min_quantity: null,
            price_list: null,
            price_list_id: null,
            updated_at: "randomString",
            variant_id: variantIds[0],
          },
          {
            amount: 130,
            created_at: "randomString",
            deleted_at: null,
            id: IdMap.getId("product-export-price_1"),
            region_id: IdMap.getId("product-export-region_1"),
            max_quantity: null,
            min_quantity: null,
            price_list: null,
            price_list_id: null,
            region: {
              id: IdMap.getId("product-export-region_3"),
              name: "denmark",
              currency_code: "dkk",
            },
            updated_at: "randomString",
            variant_id: variantIds[0],
          },
        ],
        product_id: IdMap.getId("product-export-product_1"),
        sku: "test-sku",
        title: "Test variant",
        upc: "test-upc",
        updated_at: "randomString",
        weight: null,
        width: null,
      },
    ],
    weight: null,
    width: null,
  },
  {
    sales_channels: [
      { id: IdMap.getId("sc_1"), name: "SC 1", description: "SC 1" },
      { id: IdMap.getId("sc_2"), name: "SC 2", description: "SC 2" },
    ],
    collection: {
      created_at: "randomString",
      deleted_at: null,
      handle: "test-collection2",
      id: IdMap.getId("product-export-collection_2"),
      metadata: null,
      title: "Test collection",
      updated_at: "randomString",
    },
    collection_id: "test-collection",
    created_at: "randomString",
    deleted_at: null,
    description: "test-product-description",
    discountable: true,
    external_id: null,
    handle: "test-product-product-2",
    height: null,
    hs_code: null,
    id: productIds[1],
    images: [
      {
        created_at: "randomString",
        deleted_at: null,
        id: IdMap.getId("product-export-image_2"),
        metadata: null,
        updated_at: "randomString",
        url: "test-image.png",
      },
    ],
    is_giftcard: false,
    length: null,
    material: null,
    metadata: null,
    mid_code: null,
    options: [
      {
        created_at: "randomString",
        deleted_at: null,
        id: IdMap.getId("product-export-option_2"),
        metadata: null,
        product_id: productIds[1],
        title: "test-option",
        updated_at: "randomString",
      },
    ],
    origin_country: null,
    profile_id: IdMap.getId("product-export-profile_2"),
    profile: {
      id: IdMap.getId("product-export-profile_2"),
      name: "profile_2",
      type: "profile_type_2",
    },
    status: "draft",
    subtitle: null,
    tags: [
      {
        created_at: "randomString",
        deleted_at: null,
        id: IdMap.getId("product-export-tag_2"),
        metadata: null,
        updated_at: "randomString",
        value: "123",
      },
    ],
    thumbnail: null,
    title: "Test product",
    type: {
      created_at: "randomString",
      deleted_at: null,
      id: IdMap.getId("product-export-type_2"),
      metadata: null,
      updated_at: "randomString",
      value: "test-type",
    },
    type_id: "test-type",
    updated_at: "randomString",
    variants: [
      {
        allow_backorder: false,
        barcode: "test-barcode",
        calculated_price: null,
        created_at: "randomString",
        deleted_at: null,
        ean: "test-ean",
        height: null,
        hs_code: null,
        id: variantIds[1],
        inventory_quantity: 10,
        length: null,
        manage_inventory: true,
        material: null,
        metadata: null,
        mid_code: null,
        options: [
          {
            created_at: "randomString",
            deleted_at: null,
            id: IdMap.getId("product-export-variant_option_2"),
            metadata: null,
            option_id: IdMap.getId("product-export-option_2"),
            updated_at: "randomString",
            value: "Option 1 value 1",
            variant_id: variantIds[1],
          },
        ],
        origin_country: null,
        original_price: null,
        prices: [
          {
            amount: 110,
            created_at: "randomString",
            deleted_at: null,
            id: IdMap.getId("product-export-price_2"),
            max_quantity: null,
            min_quantity: null,
            price_list: null,
            price_list_id: null,
            region_id: IdMap.getId("product-export-region_2"),
            region: {
              id: IdMap.getId("product-export-region_2"),
              name: "Denmark",
              currency_code: "dkk",
            },
            updated_at: "randomString",
            variant_id: variantIds[1],
          },
        ],
        product_id: IdMap.getId("product-export-product_2"),
        sku: "test-sku",
        title: "Test variant",
        upc: "test-upc",
        updated_at: "randomString",
        weight: null,
        width: null,
      },
      {
        allow_backorder: false,
        barcode: "test-barcode",
        calculated_price: null,
        created_at: "randomString",
        deleted_at: null,
        ean: "test-ean",
        height: null,
        hs_code: null,
        id: variantIds[2],
        inventory_quantity: 10,
        length: null,
        manage_inventory: true,
        material: null,
        metadata: null,
        mid_code: null,
        options: [
          {
            created_at: "randomString",
            deleted_at: null,
            id: IdMap.getId("product-export-variant_option_2"),
            metadata: null,
            option_id: IdMap.getId("product-export-option_2"),
            updated_at: "randomString",
            value: "Option 1 Value 1",
            variant_id: variantIds[2],
          },
        ],
        origin_country: null,
        original_price: null,
        prices: [
          {
            amount: 120,
            created_at: "randomString",
            currency_code: "usd",
            deleted_at: null,
            id: IdMap.getId("product-export-price_2"),
            max_quantity: null,
            min_quantity: null,
            price_list: null,
            price_list_id: null,
            region_id: IdMap.getId("product-export-region_1"),
            updated_at: "randomString",
            variant_id: variantIds[2],
          },
        ],
        product_id: productIds[1],
        sku: "test-sku",
        title: "Test variant",
        upc: "test-upc",
        updated_at: "randomString",
        weight: null,
        width: null,
      },
    ],
    weight: null,
    width: null,
  },
]
