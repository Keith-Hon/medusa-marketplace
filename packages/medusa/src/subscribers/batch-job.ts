import BatchJobService from "../services/batch-job"
import EventBusService from "../services/event-bus"
import { StrategyResolverService } from "../services"

type InjectedDependencies = {
  eventBusService: EventBusService
  batchJobService: BatchJobService
  strategyResolverService: StrategyResolverService
}

class BatchJobSubscriber {
  private readonly eventBusService_: EventBusService
  private readonly batchJobService_: BatchJobService
  private readonly strategyResolver_: StrategyResolverService

  constructor({
    eventBusService,
    batchJobService,
    strategyResolverService,
  }: InjectedDependencies) {
    this.eventBusService_ = eventBusService
    this.batchJobService_ = batchJobService
    this.strategyResolver_ = strategyResolverService

    this.eventBusService_
      .subscribe(BatchJobService.Events.CREATED, this.preProcessBatchJob)
      .subscribe(BatchJobService.Events.CONFIRMED, this.processBatchJob)
  }

  preProcessBatchJob = async (data): Promise<void> => {
    const batchJob = await this.batchJobService_.retrieve(data.id)

    const batchJobStrategy = this.strategyResolver_.resolveBatchJobByType(
      batchJob.type
    )

    await batchJobStrategy.preProcessBatchJob(batchJob.id)

    await this.batchJobService_.setPreProcessingDone(batchJob.id)
  }

  processBatchJob = async (data): Promise<void> => {
    const batchJob = await this.batchJobService_.retrieve(data.id)

    const batchJobStrategy = this.strategyResolver_.resolveBatchJobByType(
      batchJob.type
    )

    await this.batchJobService_.setProcessing(batchJob.id)

    await batchJobStrategy.processJob(batchJob.id)

    await this.batchJobService_.complete(batchJob.id)
  }
}

export default BatchJobSubscriber
