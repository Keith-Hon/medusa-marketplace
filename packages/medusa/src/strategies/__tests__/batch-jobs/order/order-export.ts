import OrderExportStrategy from "../../../batch-jobs/order/export"
import { IdMap, MockManager } from "medusa-test-utils"
import { User } from "../../../../models"
import { BatchJobStatus } from "../../../../types/batch-job"
import { ordersToExport } from "../../../__fixtures__/order-export-data"
import { FlagRouter } from "../../../../utils/flag-router";
import SalesChannelFeatureFlag from "../../../../loaders/feature-flags/sales-channels";

const orderServiceMock = {
  withTransaction: function (): any {
    return this
  },
  listAndCount: jest
    .fn()
    .mockImplementation(() =>
      Promise.resolve([ordersToExport, ordersToExport.length])
    ),
  list: jest.fn().mockImplementation(() => Promise.resolve(ordersToExport)),
}
const orderServiceWithoutDataMock = {
  ...orderServiceMock,
  listAndCount: jest
    .fn()
    .mockImplementation(() =>
      Promise.resolve([[], 0])
    ),
  list: jest.fn().mockImplementation(() => Promise.resolve([])),
}

describe("Order export strategy", () => {
  const outputDataStorage: string[] = []

  let fakeJob = {
    id: IdMap.getId("order-export-job"),
    type: "order-export",
    context: {
      params: {},
      list_config: {
        select: [
          "id",
          "display_id",
          "status",
          "created_at",
          "fulfillment_status",
          "payment_status",
          "subtotal",
          "shipping_total",
          "discount_total",
          "gift_card_total",
          "refunded_total",
          "tax_total",
          "total",
          "currency_code",
          "region_id",
        ],
        relations: ["customer", "shipping_address"],
      },
    },
    created_by: IdMap.getId("order-export-job-creator"),
    created_by_user: {} as User,
    result: {},
    dry_run: false,
    status: BatchJobStatus.PROCESSING,
  }

  const fileServiceMock = {
    delete: jest.fn(),
    withTransaction: function () {
      return this
    },
    getUploadStreamDescriptor: jest.fn().mockImplementation(() => {
      return Promise.resolve({
        writeStream: {
          write: (data: string) => {
            outputDataStorage.push(data)
          },
          end: () => void 0,
        },
        promise: Promise.resolve(),
        fileKey: "order-export.csv",
      })
    }),
  }
  const batchJobServiceMock = {
    withTransaction: function (): any {
      return this
    },
    update: jest.fn().mockImplementation(async (job, data) => {
      fakeJob = {
        ...fakeJob,
        ...data,
        context: { ...fakeJob?.context, ...data?.context },
        result: { ...fakeJob?.result, ...data?.result }
      }
      return Promise.resolve(fakeJob)
    }),
    complete: jest.fn().mockImplementation(async () => {
      fakeJob.status = BatchJobStatus.COMPLETED
      return fakeJob
    }),
    retrieve: jest.fn().mockImplementation(async () => {
      return fakeJob
    }),
  }

  const orderExportStrategy = new OrderExportStrategy({
    batchJobService: batchJobServiceMock as any,
    fileService: fileServiceMock as any,
    orderService: orderServiceMock as any,
    manager: MockManager,
    featureFlagRouter: new FlagRouter({}),
  })

  it("Should generate header as template", async () => {
    const template = await orderExportStrategy.buildTemplate()
    expect(template.split(";")).toEqual([
      "Order_ID",
      "Display_ID",
      "Order status",
      "Date",
      "Customer First name",
      "Customer Last name",
      "Customer Email",
      "Customer ID",
      "Shipping Address 1",
      "Shipping Address 2",
      "Shipping Country Code",
      "Shipping City",
      "Shipping Postal Code",
      "Shipping Region ID",
      "Fulfillment Status",
      "Payment Status",
      "Subtotal",
      "Shipping Total",
      "Discount Total",
      "Gift Card Total",
      "Refunded Total",
      "Tax Total",
      "Total",
      "Currency Code\r\n",
    ])
  })

  it("should process the batch job and generate the appropriate output", async () => {
    await orderExportStrategy.processJob(fakeJob.id)

    expect(outputDataStorage).toMatchSnapshot()
    expect((fakeJob.result as any).file_key).toBeDefined()
  })

  it("should always provide a file_key even with no data", async () => {
    const orderExportStrategy = new OrderExportStrategy({
      batchJobService: batchJobServiceMock as any,
      fileService: fileServiceMock as any,
      orderService: orderServiceWithoutDataMock as any,
      manager: MockManager,
      featureFlagRouter: new FlagRouter({}),
    })

    await orderExportStrategy.processJob(fakeJob.id)

    expect((fakeJob.result as any).file_key).toBeDefined()
  })
})

describe("Order export strategy with sales channel", () => {
  const outputDataStorage: string[] = []

  let fakeJob = {
    id: IdMap.getId("order-export-job"),
    type: "order-export",
    context: {
      params: {},
      list_config: {
        select: [
          "id",
          "display_id",
          "status",
          "created_at",
          "fulfillment_status",
          "payment_status",
          "subtotal",
          "shipping_total",
          "discount_total",
          "gift_card_total",
          "refunded_total",
          "tax_total",
          "total",
          "currency_code",
          "region_id",
        ],
        relations: ["customer", "shipping_address", "sales_channel"],
      },
    },
    created_by: IdMap.getId("order-export-job-creator"),
    created_by_user: {} as User,
    result: {},
    dry_run: false,
    status: BatchJobStatus.PROCESSING,
  }

  const fileServiceMock = {
    delete: jest.fn(),
    withTransaction: function () {
      return this
    },
    getUploadStreamDescriptor: jest.fn().mockImplementation(() => {
      return Promise.resolve({
        writeStream: {
          write: (data: string) => {
            outputDataStorage.push(data)
          },
          end: () => void 0,
        },
        promise: Promise.resolve(),
        fileKey: "order-export.csv",
      })
    }),
  }
  const batchJobServiceMock = {
    withTransaction: function (): any {
      return this
    },
    update: jest.fn().mockImplementation(async (job, data) => {
      fakeJob = {
        ...fakeJob,
        ...data,
        context: { ...fakeJob?.context, ...data?.context },
        result: { ...fakeJob?.result, ...data?.result }
      }
      return Promise.resolve(fakeJob)
    }),
    complete: jest.fn().mockImplementation(async () => {
      fakeJob.status = BatchJobStatus.COMPLETED
      return fakeJob
    }),
    retrieve: jest.fn().mockImplementation(async () => {
      return fakeJob
    }),
  }

  const orderExportStrategy = new OrderExportStrategy({
    batchJobService: batchJobServiceMock as any,
    fileService: fileServiceMock as any,
    orderService: orderServiceMock as any,
    manager: MockManager,
    featureFlagRouter: new FlagRouter({
      [SalesChannelFeatureFlag.key]: true,
    }),
  })

  it("Should generate header as template", async () => {
    const template = await orderExportStrategy.buildTemplate()
    expect(template.split(";")).toEqual([
      "Order_ID",
      "Display_ID",
      "Order status",
      "Date",
      "Customer First name",
      "Customer Last name",
      "Customer Email",
      "Customer ID",
      "Shipping Address 1",
      "Shipping Address 2",
      "Shipping Country Code",
      "Shipping City",
      "Shipping Postal Code",
      "Shipping Region ID",
      "Fulfillment Status",
      "Payment Status",
      "Subtotal",
      "Shipping Total",
      "Discount Total",
      "Gift Card Total",
      "Refunded Total",
      "Tax Total",
      "Total",
      "Currency Code",
      "Sales channel name",
      "Sales channel description\r\n",
    ])
  })

  it("should process the batch job and generate the appropriate output", async () => {
    await orderExportStrategy.processJob(fakeJob.id)

    expect(outputDataStorage).toMatchSnapshot()
    expect((fakeJob.result as any).file_key).toBeDefined()
  })

  it("should always provide a file_key even with no data", async () => {
    const orderExportStrategy = new OrderExportStrategy({
      batchJobService: batchJobServiceMock as any,
      fileService: fileServiceMock as any,
      orderService: orderServiceWithoutDataMock as any,
      manager: MockManager,
      featureFlagRouter: new FlagRouter({}),
    })

    await orderExportStrategy.processJob(fakeJob.id)

    expect((fakeJob.result as any).file_key).toBeDefined()
  })
})