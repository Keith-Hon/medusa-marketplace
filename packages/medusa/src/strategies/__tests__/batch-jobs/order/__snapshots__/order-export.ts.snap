// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Order export strategy should process the batch job and generate the appropriate output 1`] = `
Array [
  "Order_ID;Display_ID;Order status;Date;Customer First name;Customer Last name;Customer Email;Customer ID;Shipping Address 1;Shipping Address 2;Shipping Country Code;Shipping City;Shipping Postal Code;Shipping Region ID;Fulfillment Status;Payment Status;Subtotal;Shipping Total;Discount Total;Gift Card Total;Refunded Total;Tax Total;Total;Currency Code
",
  "order_1;123;pending;Tue, 01 Jan 2019 00:00:00 GMT;John;Doe;John@Doe.com;customer_1;123 Main St;;US;New York;10001;region_1;partially_fulfilled;captured;10;10;0;0;0;5;25;usd
",
  "order_2;124;completed;Tue, 01 Jan 2019 00:00:00 GMT;Jane;Doe;Jane@Doe.com;customer_2;Hovedgaden 1;;DK;Copenhagen;1150;region_2;fulfilled;captured;125;10;0;0;0;0;135;eur
",
]
`;

exports[`Order export strategy with sales channel should process the batch job and generate the appropriate output 1`] = `
Array [
  "Order_ID;Display_ID;Order status;Date;Customer First name;Customer Last name;Customer Email;Customer ID;Shipping Address 1;Shipping Address 2;Shipping Country Code;Shipping City;Shipping Postal Code;Shipping Region ID;Fulfillment Status;Payment Status;Subtotal;Shipping Total;Discount Total;Gift Card Total;Refunded Total;Tax Total;Total;Currency Code;Sales channel name;Sales channel description
",
  "order_1;123;pending;Tue, 01 Jan 2019 00:00:00 GMT;John;Doe;John@Doe.com;customer_1;123 Main St;;US;New York;10001;region_1;partially_fulfilled;captured;10;10;0;0;0;5;25;usd;SC 1;SC 1
",
  "order_2;124;completed;Tue, 01 Jan 2019 00:00:00 GMT;Jane;Doe;Jane@Doe.com;customer_2;Hovedgaden 1;;DK;Copenhagen;1150;region_2;fulfilled;captured;125;10;0;0;0;0;135;eur;SC 2;SC 2
",
]
`;
